esp-hal-common/src/soc/esp32s3/psram.rs

```diff
@@ -727,3 +740,1023 @@ pub(crate) mod utils {
+        // Disable the variable dummy mode when doing timing tuning
+        clear_peri_reg_mask(SPI1_MEM_DDR_REG, SPI_MEM_SPI_FMEM_VAR_DUMMY);
+        // STR /DTR mode setting is done every time when
+        // `esp_rom_opiflash_exec_cmd` is called
+        //
+        // Add any registers that are not set in ROM SPI flash functions here in
+        // the future
+    }
+
+    #[ram]
+    fn mspi_pin_init() {
+        unsafe {
+            esp_rom_opiflash_pin_config();
+        }
+        spi_timing_set_pin_drive_strength();
+        // Set F4R4 board pin drive strength. TODO: IDF-3663
+    }
+
+    #[ram]
+    fn spi_timing_set_pin_drive_strength() {
+        // For now, set them all to 3. Need to check after QVL test results are out.
+        // TODO: IDF-3663 Set default clk
+        set_peri_reg_mask(SPI0_MEM_DATE_REG, SPI_MEM_SPICLK_PAD_DRV_CTL_EN);
+        set_peri_reg_bits(
+            SPI0_MEM_DATE_REG,
+            SPI_MEM_SPI_SMEM_SPICLK_FUN_DRV,
+            3,
+            SPI_MEM_SPI_SMEM_SPICLK_FUN_DRV_S,
+        );
+        set_peri_reg_bits(
+            SPI0_MEM_DATE_REG,
+            SPI_MEM_SPI_FMEM_SPICLK_FUN_DRV,
+            3,
+            SPI_MEM_SPI_FMEM_SPICLK_FUN_DRV_S,
+        );
+        // Set default mspi d0 ~ d7, dqs pin drive strength
+        let pins = &[27usize, 28, 31, 32, 33, 34, 35, 36, 37];
+        for pin in pins {
+            unsafe {
+                let iomux = &*esp32s3::IO_MUX::PTR;
+                iomux.gpio[*pin].modify(|_, w| w.fun_drv().variant(3))
+            }
+        }
+    }
+
+    const SPI_MEM_SPICLK_PAD_DRV_CTL_EN: u32 = 1 << 4;
+    const SPI_MEM_SPI_FMEM_SPICLK_FUN_DRV: u32 = 0x00000003;
+    const SPI_MEM_SPI_FMEM_SPICLK_FUN_DRV_S: u32 = 2;
+
+    fn spi_timing_enter_mspi_low_speed_mode(control_spi1: bool) {
+        // Here we are going to slow the SPI1 frequency to 20Mhz, so we need to
+        // set SPI1 din_num and din_mode regs.
+        //
+        // Because SPI0 and SPI1 share the din_num and din_mode regs, so if we
+        // clear SPI1 din_num and din_mode to 0, if the SPI0 flash
+        // module clock is still in high freq, it may not work correctly.
+        //
+        // Therefore, here we need to slow both the SPI0 and SPI1 and related
+        // timing tuning regs to 20Mhz configuration.
+        // Switch SPI1 and SPI0 clock as 20MHz, set its SPIMEM core clock as 80M and set
+        // clock division as 4
+        spi0_timing_config_set_core_clock(SpiTimingConfigCoreClock::SpiTimingConfigCoreClock80m); // SPI0 and SPI1 share the register for core clock. So we only set SPI0 here.
+        spi0_timing_config_set_flash_clock(4);
+        if control_spi1 {
+            // After tuning, won't touch SPI1 again
+            spi1_timing_config_set_flash_clock(4);
+        }
+
+        // Set PSRAM module clock
+        spi0_timing_config_set_psram_clock(4);
+
+        // #if SPI_TIMING_FLASH_NEEDS_TUNING || SPI_TIMING_PSRAM_NEEDS_TUNING
```

Comment:
Can this be removed?

Reply:
it's meant as a reminder for adding that functionality later